ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/OffsetBuilder.scala

```diff
@@ -4,72 +4,101 @@
-    val (highest, middle, _) = split(offset)
-    fromLong(highest, middle)
+    val (version, highest, middle, _) = split(offset)
+    of(version, highest, middle)
   }
 
   def setMiddleIndex(offset: Offset, middle: Int): Offset = {
-    val (highest, _, lowest) = split(offset)
-    fromLong(highest, middle, lowest)
+    val (version, highest, _, lowest) = split(offset)
+    of(version, highest, middle, lowest)
   }
 
   def setLowestIndex(offset: Offset, lowest: Int): Offset = {
-    val (highest, middle, _) = split(offset)
-    fromLong(highest, middle, lowest)
+    val (version, highest, middle, _) = split(offset)
+    of(version, highest, middle, lowest)
   }
 
-  def fromLong(first: Long, second: Int = 0, third: Int = 0): Offset = {
+  def of(version: Byte, first: Long, second: Int = 0, third: Int = 0): Offset = {
     val bytes = new ByteArrayOutputStream
     val stream = new DataOutputStream(bytes)
-    stream.writeLong(first)
+    stream.writeByte(version.toInt)
+    writeSevenByteLong(first, stream)
     stream.writeInt(second)
     stream.writeInt(third)
     Offset.fromByteArray(bytes.toByteArray)
   }
 
+  def version(offset: Offset): Byte = split(offset)._1
+
   // `highestIndex` is used a lot, so it's worth optimizing a little rather than reusing `split`.
   def highestIndex(offset: Offset): Long = {
     val stream = new DataInputStream(new ByteArrayInputStream(offset.toByteArray))
-    stream.readLong()
+    stream.skipBytes(1) // skip version
+    readSevenByteLong(stream)
   }
 
-  def middleIndex(offset: Offset): Int = split(offset)._2
+  def middleIndex(offset: Offset): Int = split(offset)._3
 
-  def lowestIndex(offset: Offset): Int = split(offset)._3
+  def lowestIndex(offset: Offset): Int = split(offset)._4
 
-  def split(offset: Offset): (Long, Int, Int) = {
+  def split(offset: Offset): (Byte, Long, Int, Int) = {
     val stream = new DataInputStream(new ByteArrayInputStream(offset.toByteArray))
-    val highest = stream.readLong()
+    val version = stream.readByte()
+    val highest = readSevenByteLong(stream)
     val middle = stream.readInt()
     val lowest = stream.readInt()
-    (highest, middle, lowest)
+    (version, highest, middle, lowest)
+  }
+
+  private def readSevenByteLong(stream: DataInputStream): Long = {
+    // mostly copied from DataInputStream.readLong
+    val readBuffer = new Array[Byte](HighestIndexSize)
+    stream.readFully(readBuffer)
+    (readBuffer(0).toLong << 48) +
+      ((readBuffer(1) & 255).toLong << 40) +
+      ((readBuffer(2) & 255).toLong << 32) +
+      ((readBuffer(3) & 255).toLong << 24) +
+      ((readBuffer(4) & 255) << 16) +
+      ((readBuffer(5) & 255) << 8) +
+      ((readBuffer(6) & 255) << 0)
+  }
+
+  private def writeSevenByteLong(first: Long, stream: DataOutputStream): Unit = {
+    val buffer = ByteBuffer.allocate(8)
+    buffer.putLong(first << 8) // shift to preserve the sign bit
+    stream.write(buffer.array(), 0, HighestIndexSize)
```

Comment:
I'm not understanding the shift, and why you want to preserve the sign bit. Why not something like:

```suggestion
    buffer.putLong(first)
    stream.write(buffer.array(), 1, HighestIndexSize + 1)
```

If you preserve the sign bit, won't shifting overwrite bit 2? Surely if the sign bit is set, we should throw or something?

Reply:
Initially, I wasn't sure if we allow negative numbers. The tests were checking them. I had an impression that there can be a situation where the byte array looks like this: [-1, 2, 3, 4, 5, 6, 7, 8] and we have to shift to make it look like [-2, 3, 4, 5, 6, 7, 8, 0]. It may be total BS, so if we don't support negative numbers, let's not bother with that. `+ 1` is not needed, as it is a length.